torch/csrc/autograd/profiler_kineto.cpp

```diff
@@ -138,16 +138,28 @@ struct TORCH_API KinetoThreadLocalState : public ProfilerThreadLocalState {
   void finalizeCPUTrace() {
     TORCH_INTERNAL_ASSERT(cpu_trace->activities.size() == kineto_events_.size());
     for (size_t idx = 0; idx < cpu_trace->activities.size(); ++idx) {
-      if (kineto_events_[idx].hasShapes()) {
-        cpu_trace->activities[idx].addMetadata("Input Dims", shapesToStr(kineto_events_[idx].shapes()));
-      }
+      auto& kineto_event = kineto_events_[idx];
+      auto& activity = cpu_trace->activities[idx];
 
-      if (kineto_events_[idx].hasStack()) {
-        cpu_trace->activities[idx].addMetadata("Call stack", stacksToStr(kineto_events_[idx].stack()));
+      if (kineto_event.hasShapes()) {
+        activity.addMetadata("Input Dims", shapesToStr(kineto_event.shapes()));
+      }
+      if (kineto_event.hasStack()) {
+        activity.addMetadata("Call stack", stacksToStr(kineto_event.stack()));
+      }
+      if (kineto_event.hasTypes()) {
+        activity.addMetadata("Input type", dtypesToStr(kineto_event.dtypes()));
       }
 
-      if (kineto_events_[idx].hasTypes()) {
-        cpu_trace->activities[idx].addMetadata("Input type", dtypesToStr(kineto_events_[idx].dtypes()));
+      // add information about an associated forward op, if a sequence number
+      // is available (e.g. during training)
+      if (kineto_event.sequenceNr() >= 0) {
```

Comment:
May I ask what is the sequence number here, and how to associate the number and the thread id to the forward op?

Reply:
when autograd is enabled, we mark  forward and backward nodes with these sequence numbers (same number for the fwd and the associated backward op), you can find how this is done by grepping for `at::sequence_number::` ; in addition to this number we also save `fwdThreadId` - thread id of the forward op, because the sequence number itself is thread local